### PR TITLE
Introducing testing conventions task

### PR DESCRIPTION
**Pull request by alpar-t**
This is in the context of #33524 and #34820 and an annoyance I had with running these checks as a separate process. 

This PR introduces a new task meant to replace  `NamingConventionsCheck` and `NamingConventionsTask`. 
In the interest of keeping the PR small it does not do so yet, but the goal is to have all the checks implemented here. 
These checks will be added as follow ups and we will eventually replace the old tasks with the new porting the existing tests to the new task.

There are small changes in the way the checks are done, like looking for method naming connections and methods annotated with `@Test`.  

The advantage of doing this in Gradle is speed and the ability to gather insights. 
This will allow us to do things like  auto disable testing tasks if all tests are muted. 

We call the same reflection methods multiple times, like getting the methods of a class multiple times to keep what's being checked for more readable.

### CODE REVIEW

**Comment by atorok** [comment]
@elasticmachine run gradle tests 1

**Comment by atorok** [comment]
@elasticmachine run gradle build tests 1

**Comment by atorok** [comment]
@elasticmachine run gradle build tests 1

File: buildSrc/src/main/java/org/elasticsearch/gradle/precommit/TestingConventionsTasks.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class TestingConventionsTasks extends DefaultTask {
+
+    private static final String TEST_CLASS_SUFIX = "Tests";
+    private static final String INTEG_TEST_CLASS_SUFIX = "IT";
+    private static final String TEST_METHOD_PREFIX = "test";
+    private final List<String> problems;
```
Comment (nik9000) [inline buildSrc/src/main/java/org/elasticsearch/gradle/precommit/TestingConventionsTasks.java]: It feels like this should be a method variable in the `test()` method.

**Review by nik9000 — CHANGES_REQUESTED** [review]
I like it! I left a minor thing. We can build on this and remove the other checks that we have as we go! wonderful.

**Review by rjernst — COMMENTED** [review]
One question. I'm also curious how long this takes vs the previous task testing conventions implementation?

**Comment by atorok** [comment]
@rjernst  `namingConventions` runs in `1m 5s` for me `testingConventions` completes in `55s` ( both are following a clean )

**Comment by atorok** [comment]
@elasticmachine run gradle build tests 1

**Comment by nik9000** [comment]
> @rjernst `namingConventions` runs in `1m 5s` for me `testingConventions` completes in `55s` ( both are following a clean )

Is that including compile or just the running the task? Does it include gradle's startup time? Or is it the runtime of running that task for all subprojects?

**Review by nik9000 — COMMENTED** [review]
I'm happy with it modulo the timing discussion.

**Review by nik9000 — APPROVED** [review]

**Review by rjernst — APPROVED** [review]
LGTM. I also think after we convert the existing namingConventions uses, we should change this task name back to `namingConventions`

**Comment by atorok** [comment]
The times are what Gradle reports after running each task after a clean has previously been ran, using the daemon and parallel. So it does include compile times as well.

**Pull request merged**